Added graphic image ZPL2 command (^GF) with ascii encoding (GRF format)

JosDeGraeve opened the pull request:
The ascii encoding is reverse engineered from the little info that
i found on this topic on page 149 and 150 on hte ZPL programming guide,
in the section documenting the ~DG Download Graphics command.

The ascii encoding is not efficient, it increases the image site by a
factor of 2.  There is an optional compression scheme which is not
implemented yet.

JosDeGraeve committed:
Added graphic image ZPL2 command (^GF) with ascii encoding (GRF format)

The ascii encoding is reverse engineered from the little info that
i found on this topic on page 149 and 150 on hte ZPL programming guide,
in the section documenting the ~DG Download Graphics command.

The ascii encoding is not efficient, it increases the image site by a
factor of 2.  There is an optional compression scheme which is not
implemented yet.

JosDeGraeve commented:
Hi @sylvain-garancher 

Just to let you know I am working on code to allow the printer_zpl2 module to print images on zpl labels.

I have coded the algorithm in a separate python script and it seems to work quite well.  I have not tested the code in this PR yet.

sylvain-garancher commented:
Great! Thanks :)
I don't have time to test for now, but I'll try to look at this PR in the next weeks.

I think it would be better to split the method in two: One to read the image file, the other to ecode the data into ascii.
This will allow someone to call the encoding method directly if the image data is not written into a file (coming from a database, for example).

JosDeGraeve committed:
Refactored the GRF encoding into a separate method
Use a PIL image as input, so the caller can decide where to source it from.

JosDeGraeve commented:
Hi, 

I've updated this PR with your comments.

The GRF encoding part is now done in a separate method, wich i agree is better and cleaner.

I have changed the input type for the print command to a PIL image, so the caller can decide how to read the image data.

There is also a PR for the OCA module here: https://github.com/OCA/report-print-send/pull/115

Regards,

JosDeGraeve commented:
FYI, We have this in production with a customer for a few weeks now.

lasley commented on zpl2.py:
```diff
@@ -425,3 +427,94 @@ def graphic_circle(self, right, down, graphic_format):
+        4 pixel per ascii character
+        """
+
+        def round_up(value, multiple):
```
I think maybe I'm misunderstanding this method. What is `multiple`? The way I'm reading the logic here, the method rounds down not up. Am I missing something?

lasley commented on zpl2.py:
```diff
@@ -425,3 +427,94 @@ def graphic_circle(self, right, down, graphic_format):
+                result += multiple - rest
+            return result
+
+        def get_byte_val(bits, word_len):
```
I think there's a built-in that can be used here, but I'm not fully understanding this method either. What is `bits` and what is `word_len`? This is returning a byte value equivalent to `sum(bits[0:word_len])`?

lasley commented on zpl2.py:
```diff
@@ -425,3 +427,94 @@ def graphic_circle(self, right, down, graphic_format):
+
+        # make sure we have a monochrome image, and load it.
+        data = pil_image.convert("1").load()
+        sizex, sizey = pil_image.size
```
Please conform to PEP-8 snake_case variable naming. `size_x, size_y`

lasley commented on zpl2.py:
```diff
@@ -425,3 +427,94 @@ def graphic_circle(self, right, down, graphic_format):
+        # The length of the first row of pixels is set in bytes and must be
+        # a multiple of 8 bytes.
+        # The zebra tools also round the x size of images to the nearest of
+        # 8 pixels.
```
This is more appropriate as a generic doc block on the method, which would reduce my above confusion.

Given this logic, shouldn't the method be named `round_to_nearest` instead? And if that is the case, you can just leverage the pre-existing round:

```python
round( number // precision ) * precision
```

Or if you are actually just rounding up:

```python
import math
math.ceil( number // precision ) * precision
```

Make sure to use the `//` instead of `/`, which forces float division in Py2. Or `from __future__ import division` I believe

lasley commented on zpl2.py:
```diff
@@ -425,3 +427,94 @@ def graphic_circle(self, right, down, graphic_format):
+        # image
+        # Fill all values with white
+        pixels = [[1 for x in range(0, row_size)] for y in range(0, sizey)]
+        for xpos in range(0, sizex):
```
Snake case all variables please

lasley commented on zpl2.py:
```diff
@@ -425,3 +427,94 @@ def graphic_circle(self, right, down, graphic_format):
+                input_value = data[xpos, ypos]
+                # PIL: pixel value 0 = white , pixel value 255 = black
+                # GRF: 1=white, 0=black
+                bit_value = input_value >> 7 ^ 1
```
Hmmm I'm confused in this math. If I'm reading this right, `input_value` can be anything from 0 to 255? And then `bit_value` can be either 0 or 1 depending on this math here?

Assuming the above, I played with the math and can't find a difference between it and the below. Am I missing something, or is it possible to actually simplify like this?


```python
bit_value = input_value < 128
```

lasley replied:
On further thought, am I also understanding correctly that all of this logic is just to convert the image from color/greyscale into an actual black and white image? Why not just use PIL?

```python
from PIL import Image 
image_file = Image.open("convert_image.png") # open color image
image_file = image_file.convert('1') # convert image to black and white
```

lasley commented on zpl2.py:
```diff
@@ -425,3 +427,94 @@ def graphic_circle(self, right, down, graphic_format):
+
+        ret = []
+        for pixel_row in pixels:
+            pixel_string = ''
```
This would be more suited (and more efficient) as a generator:

```python
row_length = len(pixel_row)  #  I would just use row_size though and save the method call to len each row iteration.
pixel_string = ''.join([
    '%X' % get_byte_val(pixel_row[x:x+4]) for x in range(0, row_length, 4)
])
```

An interesting thing to note (and why I say this is more efficient) - string concatenation is a decent amount less performant than array appends with a subsequent join. This is because an array is a mutable data structure, and a string isn't - so every time you concatenate a string, you are actually creating three strings in memory:
```
string_3 = string_1 + string_2
```

In the case of `old_string += new_string`, it's really:

```
result_string = old_string + new_string
```

lasley commented on zpl2.py:
```diff
@@ -425,3 +427,94 @@ def graphic_circle(self, right, down, graphic_format):
+
+        return ret
+
+    def graphic_image(self, right, down, pil_image):
```
IMO this method should be named differently, although I don't have any suggestions without elaboration on what it is doing. As it stands though, the method function is not totally obvious simply by the naming IMO

Garamotte replied:
For the API methods naming, I keep the names used in the [official ZPL2 documentation](https://www.zebra.com/content/dam/zebra/manuals/en-us/software/zpl-zbi2-pm-en.pdf) for each ZPL command.

This method should be named `graphic_field` to follow the logic :)

lasley replied:
Thanks for the elaboration. I agree that alignment to protocol trumps design, any day

lasley commented on zpl2.py:
```diff
@@ -425,3 +427,94 @@ def graphic_circle(self, right, down, graphic_format):
+        # by 8
+        # Should be equal to the X size of the image, divided by 8 and
+        # rounded up
+        bytes_per_row = len(ascii_data[0])/2
```
Hmmmm it looks like we're doing this dynamically, but have the `8` hardcoded in the `encode_grf_ascii` method. Am I misunderstanding the intents of the two numbers?

lasley commented on zpl2.py:
```diff
@@ -425,3 +427,94 @@ def graphic_circle(self, right, down, graphic_format):
+        bytes_per_row = len(ascii_data[0])/2
+
+        # Total of bytes is the total number of pixels in the image
+        # transferred to the printer divided by 8
```
Or maybe this is where the hardcoded 8 comes in?

lasley commented on zpl2.py:
```diff
@@ -425,3 +427,94 @@ def graphic_circle(self, right, down, graphic_format):
+        )
+
+        # Generate the ZPL II command
+        command = '{origin}{data}{stop}'.format(
```
Is there a specific reason to use `format` here? I don't personally have a preference between `string.format` and `string %`, but IMO you should make a choice and only use one. Using both within a few lines of each other is inconsistent.

Garamotte replied:
`.format` is used everywhere else, so the best is to replace the `%` by `.format`.

lasley replied:
Consistency is king

lasley reviewed (changes requested):
Thanks @JosDeGraeve - comments inline!

sylvain-garancher commented:
@JosDeGraeve @lasley 
For the image to ascii formatting, the ZPL2 doc says:
> The data parameter is a string of hexadecimal numbers sent as a representation of the graphic 
image. Each hexadecimal character represents a horizontal nibble of four dots. For example, if the 
first four dots of the graphic image are white and the next four black, the dot-by-dot binary code is 00001111. The hexadecimal representation of this binary value is 0F. The entire graphic image is 
coded in this way, and the complete graphic image is sent as one continuous string of hexadecimal values.

The work of the `encode_grf_ascii` method can be done using `PIL` and `binascii` modules.
I propose to replace the `graphic_image` method (to be renamed to `graphic_field`) by:
```python
import binascii
from PIL import Image, ImageOps

[...]

    def graphic_image(self, right, down, pil_image):
        """ Encode a PIL image into an ASCII string suitable for ZPL printers """
        width, height = pil_image.size
        rounded_width = math.ceil(width / 8.) * 8

        # Transform the image :
        #   - Invert the colors (PIL uses 0 for black, ZPL uses 0 for white)
        #   - Convert to monochrome in case it is not already
        #   - Round the width to a multiple of 8 because ZPL needs an integer
        #     count of bytes per line (each pixel is a bit)
        pil_image = ImageOps.invert(pil_image).convert('1'). \
            pil_image.crop((0, 0, rounded_width, height))

        # Convert the image to a two-character hexadecimal values string
        ascii_data = binascii.hexlify(pil_image.tobytes()).upper()

        # Each byte is composed of two characters
        bytes_per_row = len(ascii_data) / 2
        total_bytes = bytes_per_row * height

        graphic_image_command = (
            '^GFA,{total_bytes},{total_bytes},{bytes_per_row},{ascii_data}'.format(
                total_bytes=total_bytes,
                bytes_per_row=bytes_per_row,
                ascii_data=ascii_data,
            )
        )

        # Generate the ZPL II command
        command = '{origin}{data}{stop}'.format(
            origin=self._field_origin(right, down),
            data=graphic_image_command,
            stop=self._field_data_stop(),
        )
        self._write_command(command)
```

sylvain-garancher commented:
I merged my version in master after some tests, as it worked with several images and is simpler to read.

I also made the PIL dependency optional, because someone could want to use the zpl2 module, but not graphics.

Thanks a lot for the feature proposal and for beggining the work @JosDeGraeve :)
